### PR TITLE
fit_sngls_by_template analysis time output

### PR DESCRIPTION
**Pull request by GarethDaviesGW**
Adding `analysis_time` output to `fit_sngls_by_template`

This will be needed for calculation of the new multiiforanking statistic (noise rate for each template = number of triggers/analysis time)

The analysis time is calculated through the trigger file segments and vetos as used in the fitting already

### CODE REVIEW

**Comment by GarethDaviesGW** [comment]
analysis_time output value has been tested by comparison with pycbc_page_segtable output, and found to be within a few %

count_in_template has been tested by manually counting number in a template through sum(trigger_file['L1/template_id'][:] == 123456) and found to be the same

**Comment by GarethDaviesGW** [comment]
Another test for total count in template is that similar templates have similar numbers of counts which we can see here (template_duration smoothed): https://ldas-jobs.ligo.caltech.edu/~gdavies/analyses/multiifodev/O3dev_analysis_22_multiifotest_no_inj_strip/9._manual_run/L1-FIT_OVER_PARAM_COUNT_TOTAL_VS_DURATION.png

**Comment by tdent** [comment]
OK.  Actually, I think it makes sense to combine this with changes to fit_over_param, so can you commit those and push them to this PR too?

File: bin/hdfcoinc/pycbc_fit_sngls_by_template [hunk 1]
```diff
@@ -122,6 +122,20 @@ trigf = h5py.File(args.trigger_file, 'r')
...
 templatef = h5py.File(args.bank_file, 'r')
 
+# Count the number of triggers in each template, for calulation of the true noise rate
+logging.info('Counting number of triggers in each template')
+tb = trigf[args.ifo+'/template_boundaries'][:]
+hash_sort = np.argsort(templatef['template_hash'][:])
+tb_sorted = tb[hash_sort]
+# calculate the differences between the boundary indices, 
+# in order to get the number in each template
+count_in_template = tb_sorted[1:] - tb_sorted[:-1]
```
Comment (tdent) [inline bin/hdfcoinc/pycbc_fit_sngls_by_template]: can you use numpy.diff() for this?

File: bin/hdfcoinc/pycbc_fit_sngls_by_template [hunk 2]
```diff
@@ -122,6 +122,20 @@ trigf = h5py.File(args.trigger_file, 'r')
...
+#Add on the last template
+count_in_template = np.append(count_in_template,
+                               [len(trigf[args.ifo+'/template_id'][:]) - tb_sorted[-1]])
+
+tid_sort = np.argsort(hash_sort)
```
Comment (tdent) [inline bin/hdfcoinc/pycbc_fit_sngls_by_template]: this is confusing to me - hash sort is already an argsort, from line 128, what is being done here?

**Comment by tdent** [comment]
@ahnitz it would be helpful if you could look over the template hash/id usage here

File: bin/hdfcoinc/pycbc_fit_sngls_by_template (hunk 2 above)
Comment (GarethDaviesGW) [reply bin/hdfcoinc/pycbc_fit_sngls_by_template]: by sorting the hash_sort we get an order in template id (the original order) which will put count_in_template back into the order of the original list

File: bin/hdfcoinc/pycbc_fit_sngls_by_template (hunk 1 above)
Comment (GarethDaviesGW) [reply bin/hdfcoinc/pycbc_fit_sngls_by_template]: yes - I had just used this form as it is what was used elsewhere (https://github.com/gwastro/pycbc/blob/8fe2b5ed427a33b74ca862de85acee5336c323f7/bin/hdfcoinc/pycbc_coinc_mergetrigs#L140). I will change to diff though

File: bin/hdfcoinc/pycbc_fit_sngls_by_template (hunk 2 above)
Comment (GarethDaviesGW) [reply bin/hdfcoinc/pycbc_fit_sngls_by_template]: so for any list b:
`all(b[np.argsort(b)][np.argsort(np.argsort(b))]==b)` will be True

Comment (tdent) [reply bin/hdfcoinc/pycbc_fit_sngls_by_template]: ok, if you think this is the simplest way to achieve that can you write a comment to explain it ..

**Comment by tdent** [comment]
@GarethDaviesGW can you give the command line for the analysis time test, the resulting output value and the numbers shown on the result page segtables (including time lost to vetoes)?  want to understand how close they are.

**Comment by GarethDaviesGW** [comment]
> @GarethDaviesGW can you give the command line for the analysis time test, the resulting output value and the numbers shown on the result page segtables (including time lost to vetoes)? want to understand how close they are.

So it is actually closer than I thought as the couple of tests I have done have shown it to be exact (I had forgotten to take off the veto time)

Command lines used:

```pycbc_fit_sngls_by_template --trigger-file full_data/L1-HDF_TRIGGER_MERGE_FULL_DATA-1187312718-428100.hdf --veto-file results/1._analysis_time/1.01_segment_data/L1-LIGOLW_COMBINE_SEGMENTS_CUMULATIVE_CAT_12H-1187312718-428100.xml --veto-segment-name CUMULATIVE_CAT_12H --ifo L1 --sngl-stat new_snr --fit-function exponential --stat-threshold 6 --f-lower 20 --min-duration 10 --bank-file bank/H1L1V1-BANK2HDF-1187312718-428100.hdf --output inj_coinc_manual/test_fit_sngls_out.hdf --template-fraction-range 0/20```


In ipython, test the output:
```In [1]: import h5py; fbt = h5py.File('test_fit_sngls_out.hdf','r')

In [2]: fbt.attrs.keys()
Out[2]: [u'ifo', u'fit_function', u'sngl_stat', u'stat_threshold', u'analysis_time']

In [3]: fbt.attrs['analysis_time']
Out[3]: 358628.0```

using #2519 segtable changes:
```pycbc_page_segtable --segment-files H1L1V1-SCIENCE-1187312718-428100.xml H1L1V1-SCIENCE_OK-1187312718-428100.xml H1L1V1-INSP_SEGMENTS-1187312718-428100.xml --segment-names SCIENCE SCIENCE_OK TRIGGERS_GENERATED --output-file H1L1V1-PAGE_SEGTABLE_SUMMARY-1187312718-428100.html --ifos H1 L1 V1```

output at:
https://ldas-jobs.ligo.caltech.edu/~gdavies/testoutput/H1L1V1-PAGE_SEGTABLE_SUMMARY-1187312718-428100.html

```pycbc_page_segtable --segment-files H1L1V1-INSP_SEGMENTS-1187312718-428100.xml H1L1V1-CUMULATIVE_CAT_12H_VETO_SEGMENTS_CUMULATIVE_CAT_12H-1187312718-428100.xml H1-VETOTIME_CAT1-1187312718-428100.xml H1-VETOTIME_CAT2-1187312718-428100.xml L1-VETOTIME_CAT1-1187312718-428100.xml L1-VETOTIME_CAT2-1187312718-428100.xml V1-VETOTIME_CAT1-1187312718-428100.xml V1-VETOTIME_CAT2-1187312718-428100.xml  H1L1V1-SCIENCE-1187312718-428100.xml --segment-names 'TRIGGERS_GENERATED&CUMULATIVE_CAT_12H' 'SCIENCE&VETO_CAT1' --output-file H1L1-PAGE_SEGTABLE_VETO_SUMMARY-1187312718-428100.html --ifos H1 L1 V1```

Which is here:
https://ldas-jobs.ligo.caltech.edu/~gdavies/testoutput/H1L1-PAGE_SEGTABLE_VETO_SUMMARY-1187312718-428100.html

From the first table, L1 time segments `TRIGGERS_GENERATED` and removing `CUMULATIVE_CAT_12H` in L1 from the second, we get 360477-1849 = 358628, which is the same as the output from the fit_sngls 

Same result for H1:
```In [11]: fbt_h1 = h5py.File('H1-FIT_BY_TEMPLATE-1187312718-428100-TEST-OUTPUT.hdf','r')

In [12]: fbt_h1.attrs['analysis_time']
Out[12]: 341163.0```

Values from table: 342801-1638 = 341163

**Comment by tdent** [comment]
OK, this is good!  I assume exactly the same numbers would be produced by the standard page_segtable code without any alterations.  This should be visible on one of the existing output pages.

**Comment by GarethDaviesGW** [comment]
> OK, this is good! I assume exactly the same numbers would be produced by the standard page_segtable code without any alterations. This should be visible on one of the existing output pages.

Yes, the above comment has been updated (I didnt see this reply before editing) to show that the results are the same as:
https://ldas-jobs.ligo.caltech.edu/~derek.davis/cbc/O2/clean_data_runs/o2-c02-clean-analysis-22-v1.9.1/

**Comment by tdent** [comment]
@ahnitz ok, we understand the template hash/id/boundaries thing, should be OK to proceed with this without you

**Comment by GarethDaviesGW** [comment]
> Another test for total count in template is that similar templates have similar numbers of counts which we can see here (template_duration smoothed): https://ldas-jobs.ligo.caltech.edu/~gdavies/analyses/multiifodev/O3dev_analysis_22_multiifotest_no_inj_strip/9._manual_run/L1-FIT_OVER_PARAM_COUNT_TOTAL_VS_DURATION.png

Fixed a sorting issue and now this can be found here:
https://ldas-jobs.ligo.caltech.edu/~gdavies/analyses/multiifodev/O3dev_analysis_22_multiifotest_no_inj_strip/9._manual_run/L1-FIT_OVER_PARAM_COUNT_TOTAL_VS_MTOTAL-v2.png

showing much better agreement between count_above_threshold and count_in_template trends

**Pull request merged**